processing/src/main/java/org/apache/druid/segment/nested/VariantColumnSerializer.java

```diff
@@ -299,155 +301,246 @@ public void serialize(ColumnValueSelector<? extends StructuredData> selector) th
-              elementId,
-              (id) -> indexSpec.getBitmapSerdeFactory().getBitmapFactory().makeEmptyMutableBitmap()
-          ).add(rowCount);
+  /**
+   * Internal serializer used to serailize a {@link VariantColumn}. Contains the logic to write out the column to a
+   * {@link FileSmoosher}. Created by {@link VariantColumnSerializer} once it is closed for writes.
+   */
+  public static class InternalSerializer implements Serializer
+  {
+    private final String columnName;
+    private final ByteBuffer columnNameBytes;
+    private final Byte variantTypeSetByte;
+
+    private final DictionaryWriter<String> dictionaryWriter;
+    private final FixedIndexedWriter<Long> longDictionaryWriter;
+    private final FixedIndexedWriter<Double> doubleDictionaryWriter;
+    private final FrontCodedIntArrayIndexedWriter arrayDictionaryWriter;
+
+    private final SingleValueColumnarIntsSerializer encodedValueSerializer;
+    private final GenericIndexedWriter<ImmutableBitmap> bitmapIndexWriter;
+
+    private final FixedIndexedIntWriter arrayElementDictionaryWriter;
+    private final GenericIndexedWriter<ImmutableBitmap> arrayElementIndexWriter;
+    private final boolean writeDictionary;
+
+    private final DictionaryIdLookup dictionaryIdLookup;
+
+    public InternalSerializer(
+        String columnName,
+        Byte variantTypeSetByte,
+        DictionaryWriter<String> dictionaryWriter,
+        FixedIndexedWriter<Long> longDictionaryWriter,
+        FixedIndexedWriter<Double> doubleDictionaryWriter,
+        FrontCodedIntArrayIndexedWriter arrayDictionaryWriter,
+        SingleValueColumnarIntsSerializer encodedValueSerializer,
+        GenericIndexedWriter<ImmutableBitmap> bitmapIndexWriter,
+        FixedIndexedIntWriter arrayElementDictionaryWriter,
+        GenericIndexedWriter<ImmutableBitmap> arrayElementIndexWriter,
+        DictionaryIdLookup dictionaryIdLookup,
+        boolean writeDictionary
+    )
+    {
+      this.columnName = columnName;
+      this.columnNameBytes = ColumnSerializerUtils.stringToUtf8InVSizeByteBuffer(columnName);
+      this.variantTypeSetByte = variantTypeSetByte;
+      this.dictionaryWriter = dictionaryWriter;
+      this.longDictionaryWriter = longDictionaryWriter;
+      this.doubleDictionaryWriter = doubleDictionaryWriter;
+      this.arrayDictionaryWriter = arrayDictionaryWriter;
+      this.encodedValueSerializer = encodedValueSerializer;
+      this.bitmapIndexWriter = bitmapIndexWriter;
+      this.arrayElementDictionaryWriter = arrayElementDictionaryWriter;
+      this.arrayElementIndexWriter = arrayElementIndexWriter;
+      this.writeDictionary = writeDictionary;
+      this.dictionaryIdLookup = dictionaryIdLookup;
+
+      boolean[] dictionariesSorted = new boolean[]{
+          dictionaryWriter.isSorted(),
+          longDictionaryWriter.isSorted(),
+          doubleDictionaryWriter.isSorted(),
+          arrayDictionaryWriter.isSorted()
+      };
```

Comment:
these should not be checked unless `writeDictionary` is set

Reply:
Changed in the follow up PR.